### ISSUE TITLE
Feature Request: partial grades in test evaluation

### ISSUE THREAD

**Issue by manliovalenti** [eecs-autograder/autograder.io#62]
Currently, the system evaluates a student's solution by checking the return code of a test. Instructors can specify the number of points to assign if the return value matches what is expected. However, it would be helpful to introduce a feature for assigning partial grades for partially correct solutions. 

A quick solution could be adding an option in the dropdown menu to use the return value of the evaluation function as number of assigned points. This would provide instructors with greater flexibility and control over partial grading.

**Comment by james-perretta** [comment]
@manliovalenti I agree that some way of doing this would be a good feature. It's crossed my mind in the past to use the approach of interpreting the return code as the number of points--here's a rundown of what I see as the main tradeoffs:
- Easy to implement
- Relatively low effort for the instructor to incorporate
- Handling scenarios where the instructor's script terminates unexpectedly is difficult. By interpreting the return code as something other than the return code, we can't really distinguish between normal exit and unexpected errors 
- Would only allow discrete scores, or we'd have to specify something like "values 0-100 correspond to 0-10.0 in increments of 0.1"

The other main basket of potential solutions involves requiring the instructor script to print some sort of JSON-formatted (or another format) output. I'll toss out one possible data schema and a couple of rules below for discussion:

**Proposed Data Format**
```
return_code: integer
points: integer or float
```

When a test case is marked as using this format, it should print only the above data, formatted as JSON, to stdout. Any other output should be redirected to stderr. The typical output feedback options will still be available for stderr, but not for stdout. Instructors will always be able to view the contents of stdout.

**Comment by manliovalenti** [comment]
Hi @james-perretta, I'm not familiar with how the code is structured internally (so please take my comments with a grain of salt), but I do not necessarily see the two options (using the return code as grade and requiring that the scripts outputs a JSON-like string) as mutually exclusive. In the dropdown menu, one could just select whichever option they want to go with. For many purposes, outputting a JSON may be an overkill (afterall, I know many colleagues that work with the current all-or-nothing system),  but it would certainly be a more general solution. 

To me, the first option sounds more like a quick fix with a decent amount of flexibility. There should be enough room to catch unexpected errors. For example, autograder could ask the instructor to specify a particular value (say -1 or -100) as a special return code for "unexpected errors", whereas all the other values would be interpreted as points for the exercise. Yes, this is still based on the assumption that the instructor writes correct and functioning code, but, in the end, that's also the way it is now. 

For decimal grades, I don't really know if what I'm thinking of is feasible (and I am travelling at the moment, so it is a bit hard for me to check), but if autograder can deal with non discrete grades, what would be the problem in letting the instructor script returning a float? If this is possible, I'm essentially saying: let the instructors design their scripts so that points are distributes however the instructors want, and just trust them to write a working script. Trusting the instructor script seems to be a core point even in the current system, so I do not see this as a big deal. 

Requiring that the instructor's script outputs a JSON (or alike) would certiainly allow for more complex behaviors, and the infrastructure could potentially scale up better. However, this would come with an extra burden on the instructor, who would have to be sure that the output follows a specific format (this is not necessarily a big problem for json, as there are python libraries to do that, but still it could require some extra work). Also it would require some more careful design choices. I haven't thought carefully about all the possible use cases, so I don't know what would be needed/expected, but I guess this would require some non-trivial changes to the code. Again, I feel both options could live well together as separate choices in the dropdown menu (alongside with the existing ones).

**Comment by james-perretta** [comment]
@manliovalenti I definitely agree that the simplicity of just returning the desired value is appealing and that requiring JSON output would be overkill. 

What if we had the instructor script just print a string of the form "Score: XX"? I could have the system essentially grep for that string, which should eliminate all of the requirements I stated in the JSON solution. That string would theoretically be just as simple for the instructor to write as `return score;` while still preserving the return code for diagnostic purposes. 

There would be a few details to decide on, but I don't think any of these fundamentally change the amount of instructor effort, e.g.:
- Should the message be printed to stdout or stderr? (I'd think stderr)
- What happens if there are multiple matching strings in the output?
  - Could instate a rule such as "take the last match"
  - If it were helpful, we could let instructors customize the format of the message
- How flexible should the grep for "Score: XX" be?
  - Initial thought: ignore case, colon/no colon, ignore non-newline whitespace

Thoughts?

**Comment by manliovalenti** [comment]
Yes this sounds great to me, I agree that the amount of effort for the instructor is essentially the same as just returning the grade.

I don't have a strong opinion on the questions you raised, and given that I'm not familiar with how the code works nor I have a comprehensive picture of all the use cases, so I'd be happy with whatever direction is taken (I can easily adapt to that). 

For example, I was guessing that "Score: XX" should go to stdout (after all, the score is a crucial part of the output of the script, not just any informative message), but if you think stderr fits better it's fine for me! 

I think both taking the last match and letting the instructor specify the format of the score message are good options. In particular, the latter would give a lot of flexibility (I guess you'd want the instructor to provide some regex to correctly split the message and identify the grade?). Maybe we can make the message more specific, like "Final score: XX", so that it's harder to get an unwanted match?

I agree that some flexibility for the final message would be desirable (in particular, all the options you mentioned), but I don't think much more than this is necessary. Maybe we could have a "default score message" like "Score: XX" or "Final score: XX" (case insensitive, with or without :, ignore whitespace, XX either integer or float), and if more flexibility is needed, the instructor will take care of providing the regex to parse the output message so as to identify the grade?

**Comment by liamoreilly** [comment]
Hi both,

Thanks for discussing this. This is also a feature I have thought about in the past. Full disclosure, Manlio and I work together and he discussed this last week with me, where I suggested opening an issue on here as I knew @james-perretta was receptive and very helpful in the past.

I know I am a bit late to the party here with ideas. But I think whatever is implemented then the student needs feedback to know if they went right or wrong. While the score is important, it is equally important for the student to understand the score (if it is desirable by the instructor). I could imagine writing a small scoring program that outputs a load of lines as human feedback for the student and then the final scoreline.

If the way forward was JSON, then there. could be an extra field for the displayed stdout/stderr that autograder would display in the UI to the student.

If the way forward was to parse the stdout/stderr looking for a special string, then one could just output text as normal and then have a special line something like:

- score: xx
- <score: xx>
- <score xx>

I think the <> marks would mean less likely to clash with actual output. It could even be more extreme with more obscure markers (which would make the instructor's scoring program no more audacious to write). Something like:

- <!! score: xx !!>
This would almost guarantee no clashes. 

Overall I think having student feedback and the score is a must though.

I guess the interface could allow specifying a max/min score and then autograder would clamp to those values. Or we could just expect the instructor to not mess that aspect up.

**Comment by james-perretta** [comment]
@manliovalenti @liamoreilly Thanks for your thoughts on this! Updated proposal below. @MattyMay Do you have any opinions on this?


- When this feature is enabled, the test case score will be parsed from stdout, grepping for a message of the form `<!! score: XX !!>`, where XX is the score
  - Instructors must specify (via a new test case setting) the points possible for this test case
- Instructors can specify a custom regex for the score message
- Instructors can specify that the message should be parsed from stderr instead of stdout
- Instructors can use this feature by itself or in conjunction with the existing return code and output diff checks
  - Points awarded for any such checks will be added together with the parsed score. The default regex will allow negative numbers.

**Comment by MattyMay** [comment]
@james-perretta that all seems good to me. If this feature is used in conjunction with the output diff check, should instructors be required to use `stderr` for outputting partial scores? Or should the diff checker ignore those lines of output?

**Comment by james-perretta** [comment]
@MattyMay More or less--I think it would make sense to require the diff stream and score message stream to be different. I.e., you could diff stdout and print the score to stderr or vice versa, but you can't diff stdout and print the score to stdout.

**Comment by caseylouisee** [comment]
@james-perretta Just came across this request and this sounds like exactly what I'm looking for - any idea if/when this will be published?

**Comment by james-perretta** [comment]
@caseylouisee I think we'll put this on our "near future" list to start working on in the week or so. Feel free to check back in for an update in a few weeks if there's no activity here.

**Comment by manliovalenti** [comment]
@james-perretta any update on this? It'd be great if this could be working by the beginning of the next semester!

**Comment by james-perretta** [comment]
@manliovalenti Yes, we've more or less wrapped up the server-side logic and are working on the frontend. Unless we run into an unforseen problem, this should be ready to go in time for the fall semester
